src/logworker.cpp

```diff
@@ -120,6 +122,14 @@ namespace g3 {
       return std::unique_ptr<LogWorker>(new LogWorker);
    }
 
+   void LogWorker::removeWrappedSink(std::shared_ptr<g3::internal::SinkWrapper> sink) {
+      auto bg_removesink_call = [this, &sink] {
```

Comment:
Thanks @gimesketvirtadieni yes. I could see the same thing. Interesting. It's almost as if the `packaged_task` should contain the `join()` function too since otherwise you are not guaranteed resource state. 

I don't think this changes anything though as far as the code goes.  Let's go through the alternative, bottom-up
1) pass the reference --> The problem, as you already pointed out is that the lambda will be kept alive for a longer time. By passing a reference to a `std::shared_ptr` you will in these instances end up with a `dangling reference`. Although, likely, completely safe this is opening up a can of worms and goes against the usage philosophy for `share_ptrs`.     This I therefore consider the worst option. 

2. using `weak_ptr`. That's possible but I do not see a real strong reason for this. The issue seems to be a unit test issue only and one that could easily be worked around.   Once the sink is removed from the internal storage it cannot be used anymore. 

Do you see any issues with a few ms delay in removing the actual sink, apart from unit test wise?  

3. The same reasoning goes for the empty lambda. It is a crutch to fix unit test and to me, does not seem to be adding any value to the production code. 

Unless you can shed some light on why it would impact the production code I suggest you remove the dangling reference code and use plain shared_ptr.   The work around for the unit test is simple enough

```
TEST(Sink, KJELL_OneSinkRemove) {
using namespace g3;
   AtomicBoolPtr flag = make_shared < atomic<bool >> (false);
   AtomicIntPtr count = make_shared < atomic<int >> (0);
   {
      auto worker = g3::LogWorker::createLogWorker();
      auto handle = worker->addSink(std2::make_unique<ScopedSetTrue>(flag, count), &ScopedSetTrue::ReceiveMsg);
      EXPECT_FALSE(flag->load());
      EXPECT_TRUE(0 == count->load());

      LogMessagePtr message1{std2::make_unique<LogMessage>("test", 0, "test", DEBUG)};
      message1.get()->write().append("this message should trigger an atomic increment at the sink");
      worker->save(message1);
      {
         auto realSink = handle->sink();
         std::cout << "instances: " << realSink.use_count() << std::endl;
      }
      EXPECT_FALSE(flag->load());
      worker->removeSink(std::move(handle));
      size_t kMaxWaitMs = 2000;
      std::chrono::milliseconds wait{200};
      std::chrono::milliseconds maxWait{kMaxWaitMs};
      auto start = std::chrono::steady_clock::now();
      auto stopWatch = [&start]() -> size_t{
         auto end = std::chrono::steady_clock::now();
         std::chrono::duration<double> elapsedMs = end - start;
         return elapsedMs.count();
     };


      while(true) {
         auto elapsedMs = stopWatch();
         if (flag->load() || elapsedMs > kMaxWaitMs) {
           break;
         }
         std::this_thread::sleep_for(wait);
      }
      std::cout << "Max wait was: " << stopWatch() << "ms" << std::endl;
      EXPECT_TRUE(flag->load());
      EXPECT_TRUE(1 == count->load());

      LogMessagePtr message2{std2::make_unique<LogMessage>("test", 0, "test", DEBUG)};
      message2.get()->write().append("this message is issued after all sinks are removed");
      worker->save(message2);
   }
   EXPECT_TRUE(flag->load());
   EXPECT_TRUE(1 == count->load());
}

```